File: nosql-create-index-typescript/src/data-plane.ts
/**
 * Data-plane operations using @azure/cosmos and Azure OpenAI.
 *
 *   3. Verify embedding dimensions
 *   4. Insert documents from pre-vectorized data file (bulk)
 *   5. Run a vector similarity query using VectorDistance()
 */

import { CosmosClient, BulkOperationType, type Container } from "@azure/cosmos";
import { AzureOpenAI } from "openai";
import { getBearerTokenProvider, type TokenCredential } from "@azure/identity";
import { readFile } from "node:fs/promises";
import { appendFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { randomUUID } from "node:crypto";
import type { SampleConfig } from "./config.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// ---------------------------------------------------------------------------
// Client factories
// ---------------------------------------------------------------------------

/** Create a Cosmos DB data-plane client. */
export function createCosmosClient(
  credential: TokenCredential,
  endpoint: string
) {
  return new CosmosClient({
    endpoint,
    aadCredentials: credential,
  });
}

/** Create an Azure OpenAI client for embedding generation. */
export function createOpenAIClient(
  credential: TokenCredential,
  config: SampleConfig
) {
  const tokenProvider = getBearerTokenProvider(
    credential,
    "https://cognitiveservices.azure.com/.default"
  );
  return new AzureOpenAI({
    azureADTokenProvider: tokenProvider,
    endpoint: config.openai.endpoint!,
    apiVersion: config.openai.embeddingApiVersion,
  });
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Generate an embedding vector for the given text. */
async function generateEmbedding(
  openaiClient: AzureOpenAI,
  deployment: string,
  text: string
) {
  const response = await openaiClient.embeddings.create({
    model: deployment,
    input: [text],
  });
  return response.data[0].embedding;
}

// ---------------------------------------------------------------------------
// Step 3 — Verify embedding dimensions
// ---------------------------------------------------------------------------
export async function verifyEmbeddingDimensions(
  openaiClient: AzureOpenAI,
  config: SampleConfig
) {
  const embedding = await generateEmbedding(
    openaiClient,
    config.openai.embeddingDeployment,
    "dimension check"
  );
  const actual = embedding.length;

  if (actual !== config.expectedDimensions) {
    throw new Error(
      `Dimension mismatch: model produces ${actual} but container expects ${config.expectedDimensions}. ` +
        "Update EMBEDDING_DIMENSIONS and recreate the container."
    );
  }

  return actual;
}

// ---------------------------------------------------------------------------
// Step 4 — Insert documents from data file (bulk)
// ---------------------------------------------------------------------------

/**
 * Validate that all documents have a Region property and it's one of the expected regions.
 */
function validateRegionProperty(documents: any[]): void {
  const expectedRegions = new Set(["Northeast", "Midwest", "South", "West"]);
  const regionsFound = new Set<string>();

  for (let idx = 0; idx < documents.length; idx++) {
    const doc = documents[idx];
    if (!doc.Region) {
      throw new Error(
        `Document at index ${idx} (HotelId=${doc.HotelId || "unknown"}) missing Region property`
      );
    }
    const region = doc.Region;
    regionsFound.add(region);
    if (!expectedRegions.has(region)) {
      throw new Error(
        `Document at index ${idx} has unexpected Region '${region}'. Expected one of: ${Array.from(expectedRegions).join(", ")}`
      );
    }
  }

  console.log(`✓ Region validation passed. Found regions: ${Array.from(regionsFound).sort().join(", ")}`);
}

/**
 * Group documents by Region partition key.
 */
function groupByRegion(documents: any[]): Map<string, any[]> {
  const docsByRegion = new Map<string, any[]>();

  for (const doc of documents) {
    const region = doc.Region;
    if (!docsByRegion.has(region)) {
      docsByRegion.set(region, []);
    }
    docsByRegion.get(region)!.push(doc);
  }

  // Log region grouping summary
  for (const [region, docs] of Array.from(docsByRegion.entries()).sort()) {
    console.log(`  Region '${region}': ${docs.length} documents`);
  }

  return docsByRegion;
}

export async function insertDocuments(
  container: Container,
  config: SampleConfig,
  embeddingField: string,
  containerName?: string
) {
  const actualContainerName = containerName || config.cosmos.containerName;
  
  // Load pre-vectorized hotel data from JSON file
  const filePath = resolve(__dirname, "..", config.dataFile);
  const fileContent = await readFile(filePath, "utf-8");
  const data = JSON.parse(fileContent);
  console.log(`\nReading JSON file from ${filePath}`);
  console.log(`Loaded ${data.length} documents`);

  // Check if container already has documents
  const { resources: countResult } = await container.items
    .query("SELECT VALUE COUNT(1) FROM c")
    .fetchAll();

  if (countResult[0] > 0) {
    console.log(
      `  ✓ ${actualContainerName}: ${countResult[0]} documents already exist (skipped)`
    );
    return { total: data.length, inserted: 0, skipped: countResult[0] };
  }

  // Validate Region property and group by region
  validateRegionProperty(data);
  const docsByRegion = groupByRegion(data);

  console.log(`Processing by region...`);
  let inserted = 0;
  let failed = 0;
  let conflict = 0;
  let totalRU = 0;
  const errors: Array<{ doc: string; error: string; statusCode?: number }> = [];

  // Batch ingest by region (one batch per region)
  for (const [region, regionDocs] of docsByRegion) {
    const operations = regionDocs.map((item) => ({
      operationType: BulkOperationType.Create,
      resourceBody: {
        ...item,
        id: item.HotelId || randomUUID(), // Map HotelId to id (required for Cosmos bulk ops)
        // Use the pre-computed embedding from the data file (DescriptionVector)
        [embeddingField]: item.DescriptionVector || item.embedding,
      },
      partitionKey: [region],
    }));

    try {
      const response = await container.items.executeBulkOperations(operations);

      if (response) {
        for (const result of response) {
          const statusCode = result.response?.statusCode;
          const errorCode = result.error?.code;
          const ru = result.response?.requestCharge ?? 0;

          if (statusCode && Number(statusCode) >= 200 && Number(statusCode) < 300) {
            inserted++;
          } else if (Number(statusCode) === 409) {
            // 409 Conflict: document already exists
            conflict++;
            inserted++;
          } else if (result.error) {
            failed++;
            const docId = (result as any).resourceBody?.id || "unknown";
            errors.push({
              doc: docId,
              error: result.error.message || "Unknown error",
              statusCode: statusCode as number | undefined,
            });
          } else {
            inserted++;
          }
          totalRU += ru;
        }
      }
    } catch (error) {
      failed += regionDocs.length;
      errors.push({
        doc: `Region: ${region} (${regionDocs.length} docs)`,
        error: (error as Error).message || "Batch operation failed",
      });
    }
  }

  if (errors.length > 0) {
    console.error(`  ✗ Errors during ingestion:`);
    for (const err of errors.slice(0, 5)) {
      console.error(`    - ${err.doc}: ${err.error}${err.statusCode ? ` (${err.statusCode})` : ""}`);
    }
    if (errors.length > 5) {
      console.error(`    ... and ${errors.length - 5} more errors`);
    }
  }

  if (failed > 0) {
    throw new Error(
      `Batch ingestion incomplete: ${failed} documents failed, ${conflict} conflicts, ${inserted} inserted in container '${actualContainerName}'.`
    );
  }

  console.log(`  ✓ ${actualContainerName}: ${inserted} inserted (${conflict > 0 ? `${conflict} conflicts, ` : ""}${totalRU.toFixed(2)} RUs)`);
  return { total: data.length, inserted, failed };
}

// ---------------------------------------------------------------------------
// Step 5 — Vector similarity queries
// ---------------------------------------------------------------------------
export async function vectorQuery(
  container: Container,
  containerName: string,
  openaiClient: AzureOpenAI,
  config: SampleConfig
) {
  const embeddingField = config.embeddingField;
  
  // Cosmos DB SQL does not support parameter placeholders for field names,
  // so the field name is string-interpolated. Validate to prevent injection.
  if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(embeddingField)) {
    throw new Error(`Invalid embedding field name: ${embeddingField}`);
  }

  const queryText = "hotel near the ocean";
  const queryEmbedding = await generateEmbedding(
    openaiClient,
    config.openai.embeddingDeployment,
    queryText
  );

  console.log(`\nQuery: "${queryText}"`);
  console.log(`Embedding generated (${queryEmbedding.length} dimensions)`);
  console.log(`\nRunning search (top 5 results for each distance function)...`);

  // Query with 3 distance functions using VectorDistance() options parameter
  const distanceFunctions = [
    { name: "Cosine", orderDirection: "DESC" },      // Similarity: higher = better
    { name: "DotProduct", orderDirection: "DESC" },  // Similarity: higher = better
    { name: "Euclidean", orderDirection: "ASC" },    // Distance: lower = better
  ];
  const results: Array<{ container: string; metric: string; top1: string; top1Score: number; top2: string; top2Score: number; diff: number; ru: number }> = [];

  for (const distFunc of distanceFunctions) {
    const distanceFunction = distFunc.name;
    
    // Query strategy: VectorDistance automatically sorts by similarity
    // (no ORDER BY clause allowed - it's automatic and always sorts most-to-least similar)
    
    // Single-partition query for efficiency:
    // - WHERE clause filters to the configured region
    // - SDK partition key routing ensures single-partition execution
    // - Belt-and-suspenders pattern: both mechanisms for guaranteed efficiency
    const partitionKeyValue = config.partitionKeyValue;
    
    const querySpec = {
      query: `SELECT TOP 5
                c.HotelId,
                c.HotelName,
                c.Description,
                VectorDistance(c.${embeddingField}, @embedding, false, {'distanceFunction': '${distanceFunction}'}) AS SimilarityScore
              FROM c
              WHERE c.Region = @partitionKey`,
      parameters: [
        { name: "@embedding", value: queryEmbedding },
        { name: "@partitionKey", value: partitionKeyValue }
      ],
    };

    try {
      const { resources, requestCharge } = await container.items
        .query(querySpec, { partitionKey: partitionKeyValue })
        .fetchAll();

      if (resources.length >= 2) {
        const top1 = resources[0];
        const top2 = resources[1];
        results.push({
          container: containerName,
          metric: distanceFunction,
          top1: (top1.HotelName || "").substring(0, 26),
          top1Score: top1.SimilarityScore,
          top2: (top2.HotelName || "").substring(0, 26),
          top2Score: top2.SimilarityScore,
          diff: Math.abs(top1.SimilarityScore - top2.SimilarityScore),
          ru: requestCharge,
        });
      } else if (resources.length === 1) {
        const top1 = resources[0];
        results.push({
          container: containerName,
          metric: distanceFunction,
          top1: (top1.HotelName || "").substring(0, 26),
          top1Score: top1.SimilarityScore,
          top2: "",
          top2Score: 0,
          diff: 0,
          ru: requestCharge,
        });
      }
      
      console.log(`  ✓ ${containerName} queried (${requestCharge.toFixed(2)} RUs)`);
    } catch (error) {
      const errMsg = error instanceof Error ? error.message : String(error);
      console.error(`  ✗ ${distanceFunction} query failed: ${errMsg}`);
    }
  }

  return results;
}

/**
 * Delete all sample-inserted documents from a container.
 * Uses a query to find all documents and delete them in bulk by Region.
 */
export async function clearContainerData(container: Container): Promise<void> {
  const querySpec = {
    query: "SELECT c.id, c.Region FROM c",
  };

  const { resources: itemsToDelete } = await container.items.query(querySpec).fetchAll();

  if (itemsToDelete.length === 0) {
    return;
  }

  const operations = itemsToDelete.map((item: any) => ({
    operationType: BulkOperationType.Delete,
    id: item.id,
    partitionKey: [item.Region],  // Use Region as partition key
  }));

  await container.items.bulk(operations);
}

File: nosql-create-index-python/src/data_plane.py
"""Data-plane operations for the nosql-create-index-python sample."""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, Iterable, List, Sequence

from azure.cosmos import CosmosClient
from azure.cosmos.exceptions import CosmosHttpResponseError
from azure.identity import DefaultAzureCredential, get_bearer_token_provider
from openai import AzureOpenAI

from .config import SampleConfig

BATCH_SIZE = 100
FIELD_NAME_PATTERN = r"^[A-Za-z_][A-Za-z0-9_]*$"


@dataclass(frozen=True)
class IngestionSummary:
    container_name: str
    total_documents: int
    inserted_documents: int
    skipped: bool
    request_charge: float


@dataclass(frozen=True)
class QueryResult:
    hotel_id: str
    hotel_name: str
    description: str
    score: float


@dataclass(frozen=True)
class QuerySummary:
    container_name: str
    request_charge: float
    activity_id: str
    results: Sequence[QueryResult]


def create_cosmos_client(config: SampleConfig, credential: DefaultAzureCredential) -> CosmosClient:
    return CosmosClient(url=config.cosmos_endpoint, credential=credential)


def create_azure_openai_client(
    config: SampleConfig, credential: DefaultAzureCredential
) -> AzureOpenAI:
    token_provider = get_bearer_token_provider(
        credential, "https://cognitiveservices.azure.com/.default"
    )
    return AzureOpenAI(
        azure_endpoint=config.openai_embedding_endpoint,
        azure_ad_token_provider=token_provider,
        api_version=config.openai_embedding_api_version,
    )


def read_documents(data_file: Path) -> List[Dict[str, Any]]:
    with data_file.open("r", encoding="utf-8") as handle:
        payload = json.load(handle)

    if not isinstance(payload, list):
        raise ValueError("The shared hotel dataset must be a JSON array.")

    return [prepare_document(item) for item in payload]


def prepare_document(item: Dict[str, Any]) -> Dict[str, Any]:
    document = dict(item)
    document["id"] = str(item["HotelId"])
    document["PartitionKey"] = "hotels"
    return document


def generate_embedding(
    openai_client: AzureOpenAI, config: SampleConfig, text: str
) -> Sequence[float]:
    response = openai_client.embeddings.create(
        model=config.openai_embedding_deployment,
        input=[text],
    )
    return response.data[0].embedding


def verify_embedding_dimensions(
    openai_client: AzureOpenAI, config: SampleConfig
) -> Sequence[float]:
    embedding = generate_embedding(openai_client, config, "dimension check")
    actual_dimensions = len(embedding)

    if actual_dimensions != config.expected_dimensions:
        raise ValueError(
            "Embedding dimensions do not match the container definition. "
            "Expected {0}, received {1}.".format(
                config.expected_dimensions, actual_dimensions
            )
        )

    return embedding


def ingest_documents(container: Any, container_name: str, documents: Sequence[Dict[str, Any]]) -> IngestionSummary:
    existing_count = _document_count(container)
    if existing_count > 0:
        return IngestionSummary(
            container_name=container_name,
            total_documents=len(documents),
            inserted_documents=0,
            skipped=True,
            request_charge=0.0,
        )

    inserted_documents = 0
    failed_documents = 0
    total_request_charge = 0.0
    for batch in _chunked(documents, BATCH_SIZE):
        operations = [("upsert", (document,)) for document in batch]
        results = container.execute_item_batch(
            batch_operations=operations,
            partition_key="hotels",
        )
        for result in results:
            if int(result.get("statusCode", 0)) < 300:
                inserted_documents += 1
            else:
                failed_documents += 1
        total_request_charge += _request_charge(container)

    if failed_documents > 0:
        print(
            "WARNING: {0} of {1} documents failed to insert.".format(
                failed_documents, len(documents)
            )
        )
        raise RuntimeError(
            "Batch ingestion incomplete: {0} documents failed in container '{1}'.".format(
                failed_documents, container_name
            )
        )

    return IngestionSummary(
        container_name=container_name,
        total_documents=len(documents),
        inserted_documents=inserted_documents,
        skipped=False,
        request_charge=total_request_charge,
    )


def query_top_matches(
    container: Any,
    container_name: str,
    config: SampleConfig,
    query_embedding: Sequence[float],
    metric: str = "cosine",
) -> QuerySummary:
    embedding_field = validate_field_name(config.embedding_field_name)
    query_text = (
        "SELECT TOP @topK c.HotelId, c.HotelName, c.Description, "
        "VectorDistance(c.{0}, @embedding, false) AS similarityScore "
        "FROM c WHERE c.PartitionKey = @partitionKey "
        "ORDER BY VectorDistance(c.{0}, @embedding, false)"
    ).format(embedding_field)

    raw_results = list(
        container.query_items(
            query=query_text,
            parameters=[
                {"name": "@topK", "value": config.top_count},
                {"name": "@embedding", "value": list(query_embedding)},
                {"name": "@partitionKey", "value": config.partition_key_value},
            ],
            partition_key=config.partition_key_value,
        )
    )

    results = [
        QueryResult(
            hotel_id=str(item["HotelId"]),
            hotel_name=str(item["HotelName"]),
            description=str(item["Description"]),
            score=float(item["similarityScore"]),
        )
        for item in raw_results
    ]

    headers = getattr(container.client_connection, "last_response_headers", {}) or {}
    return QuerySummary(
        container_name=container_name,
        request_charge=_request_charge(container),
        activity_id=str(headers.get("x-ms-activity-id", "")),
        results=results,
    )


def print_query_summary(summary: QuerySummary, algorithm_name: str) -> None:
    print("\n=== Query results: {0} ({1}) ===".format(summary.container_name, algorithm_name))
    if summary.activity_id:
        print("Activity ID: {0}".format(summary.activity_id))
    print("Request charge: {0:.2f} RUs".format(summary.request_charge))

    for rank, result in enumerate(summary.results, start=1):
        print(
            "{0}. HotelId={1} | HotelName={2} | score={3:.4f} | Description={4}".format(
                rank,
                result.hotel_id,
                result.hotel_name,
                result.score,
                _shorten(result.description),
            )
        )


def validate_field_name(field_name: str) -> str:
    import re

    if not re.match(FIELD_NAME_PATTERN, field_name):
        raise ValueError(
            "Invalid embedding field name: {0}".format(field_name)
        )
    return field_name


def diagnose_resources(cosmos_client: CosmosClient, config: SampleConfig) -> None:
    """Diagnose whether database and containers exist."""
    print("\n=== Diagnostic Check ===")
    print("Cosmos DB Endpoint: {0}".format(config.cosmos_endpoint))
    print("Database name: {0}".format(config.database_name))
    
    try:
        database = cosmos_client.get_database_client(config.database_name)
        # Try to list containers
        containers = list(database.list_containers())
        print("✓ Database '{0}' exists".format(config.database_name))
        print("  Containers found: {0}".format(len(containers)))
        for container in containers:
            print("    - {0}".format(container["id"]))
        
        if len(containers) == 0:
            print("  ⚠ WARNING: Database exists but has NO containers.")
    except CosmosHttpResponseError as e:
        if e.status_code == 404:
            print("✗ Database '{0}' NOT FOUND (404)".format(config.database_name))
            print("  Create the database with: azd provision")
        else:
            print("✗ Error accessing database: {0} ({1})".format(e.message, e.status_code))
    except Exception as e:
        print("✗ Unexpected error checking database: {0}".format(str(e)))


def wrap_runtime_error(error: Exception) -> RuntimeError:
    if isinstance(error, CosmosHttpResponseError):
        message = getattr(error, "message", str(error)) or str(error)
        return RuntimeError(
            "Cosmos DB data-plane request failed. Verify the database, containers, "
            "and Microsoft Entra ID RBAC assignment created by the shared Bicep deployment. "
            "Original error: {0}".format(message)
        )
    return RuntimeError(str(error))


def _document_count(container: Any) -> int:
    query = "SELECT VALUE COUNT(1) FROM c WHERE c.PartitionKey = @partitionKey"
    results = list(
        container.query_items(
            query=query,
            parameters=[{"name": "@partitionKey", "value": "hotels"}],
            partition_key="hotels",
        )
    )
    return int(results[0]) if results else 0


def _chunked(values: Sequence[Dict[str, Any]], size: int) -> Iterable[Sequence[Dict[str, Any]]]:
    for index in range(0, len(values), size):
        yield values[index : index + size]


def _request_charge(container: Any) -> float:
    headers = getattr(container.client_connection, "last_response_headers", {}) or {}
    raw_value = headers.get("x-ms-request-charge", "0")
    try:
        return float(raw_value)
    except (TypeError, ValueError):
        return 0.0


def _shorten(value: str, limit: int = 110) -> str:
    if len(value) <= limit:
        return value
    return value[: limit - 3].rstrip() + "..."
